### PR TITLE
IBX-4160: Enh. custom Limitation doc and examples (4.5+)

### PR DESCRIPTION
**Pull request by adriendupuis**
| Question      | Answer
| ------------- | ---
| JIRA Ticket   | [IBX-4160](https://issues.ibexa.co/browse/IBX-4160) (a. k. a. [EZP-32155](https://issues.ibexa.co/browse/EZP-32155))
| Versions      | 4.5+
| Edition       | All

Enhance custom Limitation doc and examples.
Convert #1461 back to 4.5+ for more detailed examples.

```bash
cp -rv $DOC_DIR/code_samples/back_office/limitation/* $DXP_DIR/
tail -n +2 $DXP_DIR/config/append_to_services.yaml >> $DXP_DIR/config/services.yaml
rm $DXP_DIR/config/append_to_services.yaml
```

#### Checklist

- [ ] Text renders correctly
- [ ] Text has been checked with vale
- [ ] Description metadata is up to date
- [ ] Redirects cover removed/moved pages
- [ ] Code samples are working
- [ ] PHP code samples have been fixed with PHP CS fixer
- [ ] Added link to this PR in relevant JIRA ticket or code PR

### CODE REVIEW

File: code_samples/back_office/limitation/translations/forms.en.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+role.policy.custom_module: 'Custom module'
+role.policy.custom_module.all_functions: 'Custom module / All functions'
+role.policy.custom_module.custom_function_1: 'Custom module / Function #1'
+role.policy.custom_module.custom_function_2: 'Custom module / Function #2'
```
Comment (adriendupuis) [inline code_samples/back_office/limitation/translations/forms.en.yaml:4]: TODO: Study #2205 for an example allowing `translation:extract` to generate this file (something like `php bin/console translation:extract en --domain=forms --dir=src --output-dir=translations`)

Comment (adriendupuis) [reply code_samples/back_office/limitation/translations/forms.en.yaml:4]: Proposed in 0fae483c0a265530e72b0554faee48157eb58ad3

File: docs/permissions/custom_policies.md [hunk 2]
```diff
@@ -58,131 +60,243 @@ class MyPolicyProvider implements PolicyProviderInterface
...
-            limitation_value_templates:
-                - { template: limitation/custom_limitation_value.html.twig, priority: 0 }
+Provide translations for your custom limitation form in the `ezplatform_content_forms_policies` domain. For example, `translations/ezplatform_content_forms_policies.en.yaml`:
 
+``` yaml
+[[= include_file('code_samples/back_office/limitation/translations/ezplatform_content_forms_policies.en.yaml') =]]
 ```
 
-!!! note
+### Checking user custom Limitation
```
Comment (dabrt) [inline docs/permissions/custom_policies.md:248]: Consider rewriting all headers to remove the gerund.

Comment (adriendupuis) [reply docs/permissions/custom_policies.md]: Done in ea5117c9d97093cfafd42d56ba04a7cbb204c2f1

File: code_samples/back_office/limitation/config/append_to_services.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+services:
+
```
Comment (adriendupuis) [inline code_samples/back_office/limitation/config/append_to_services.yaml:2]:
```suggestion
services:
```

[the remark this answers is not in the document]
Comment (adriendupuis) [reply code_samples/back_office/limitation/config/append_to_services.yaml]: Removed in 299dad70d89432cebaff08344066637ab5594c4a

File: code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+use Ibexa\Contracts\Core\Repository\Values\Content\Query\CriterionInterface;
+use Ibexa\Contracts\Core\Repository\Values\User\Limitation;
+use Ibexa\Contracts\Core\Repository\Values\User\UserReference;
+use Ibexa\Contracts\Core\Repository\Values\ValueObject;
+use Ibexa\Core\Base\Exceptions\InvalidArgumentType;
+use Ibexa\Core\FieldType\ValidationError;
+
+class CustomLimitationType implements Type
+{
+    public function acceptValue(Limitation $limitationValue)
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php:18]:
```suggestion
    public function acceptValue(Limitation $limitationValue): void
```

File: code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        if (!$limitationValue instanceof CustomLimitationValue) {
+            throw new InvalidArgumentType(
+                '$limitationValue',
+                FieldGroupLimitation::class,
+                $limitationValue
+            );
+        }
+    }
+
+    public function validate(Limitation $limitationValue)
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php:29]: Or
```suggestion
    /** @return ValidationError[] */
    public function validate(Limitation $limitationValue): array
```

File: code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        if (!array_key_exists('value', $limitationValue->limitationValues)) {
+            $validationErrors[] = new ValidationError("limitationValues['value'] is missing.");
+        } elseif (!is_bool($limitationValue->limitationValues['value'])) {
+            $validationErrors[] = new ValidationError("limitationValues['value'] is not a boolean.");
+        }
+
+        return $validationErrors;
+    }
+
+    public function buildValue(array $limitationValues)
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php:41]:
```suggestion
    public function buildValue(array $limitationValues): CustomLimitationValue
```

File: code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+        return new CustomLimitationValue(['limitationValues' => ['value' => $value]]);
+    }
+
+    /**
+     * @param \Ibexa\Contracts\Core\Repository\Values\ValueObject[]|null $targets
+     *
+     * @return bool|null
+     */
+    public function evaluate(Limitation $value, UserReference $currentUser, ValueObject $object, array $targets = null)
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php:58]:
```suggestion
    public function evaluate(Limitation $value, UserReference $currentUser, ValueObject $object, array $targets = null): ?bool
```

File: code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+        return Type::ACCESS_DENIED;
+    }
+
+    public function getCriterion(Limitation $value, UserReference $currentUser): CriterionInterface
+    {
+        throw new NotImplementedException(__METHOD__);
+    }
+
+    public function valueSchema()
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php:78]:
```suggestion
    public function valueSchema(): void
```

File: code_samples/back_office/limitation/src/Security/Limitation/Mapper/CustomLimitationFormMapper.php [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+use Ibexa\AdminUi\Limitation\LimitationFormMapperInterface;
+use Ibexa\AdminUi\Translation\Extractor\LimitationTranslationExtractor;
+use Ibexa\Contracts\Core\Repository\Values\User\Limitation;
+use Symfony\Component\Form\Extension\Core\Type\CheckboxType;
+use Symfony\Component\Form\FormInterface;
+
+class CustomLimitationFormMapper implements LimitationFormMapperInterface
+{
+    public function mapLimitationForm(FormInterface $form, Limitation $data)
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply code_samples/back_office/limitation/src/Security/Limitation/Mapper/CustomLimitationFormMapper.php:15]:
```suggestion
    public function mapLimitationForm(FormInterface $form, Limitation $data): void
```

File: code_samples/back_office/limitation/src/Security/Limitation/Mapper/CustomLimitationFormMapper.php [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+            'property_path' => 'limitationValues[value]',
+        ]);
+    }
+
+    public function getFormTemplate()
+    {
+        return '@ibexadesign/limitation/custom_limitation_form.html.twig';
+    }
+
+    public function filterLimitationValues(Limitation $limitation)
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply code_samples/back_office/limitation/src/Security/Limitation/Mapper/CustomLimitationFormMapper.php:30]:
```suggestion
    public function filterLimitationValues(Limitation $limitation): void
```

File: code_samples/back_office/limitation/src/Security/Limitation/Mapper/CustomLimitationValueMapper.php [hunk 11]
```diff
@@ -0,0 +1,14 @@
+<?php declare(strict_types=1);
+
+namespace App\Security\Limitation\Mapper;
+
+use Ibexa\AdminUi\Limitation\LimitationValueMapperInterface;
+use Ibexa\Contracts\Core\Repository\Values\User\Limitation;
+
+class CustomLimitationValueMapper implements LimitationValueMapperInterface
+{
+    public function mapLimitationValue(Limitation $limitation)
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply code_samples/back_office/limitation/src/Security/Limitation/Mapper/CustomLimitationValueMapper.php:12]:
```suggestion
    public function mapLimitationValue(Limitation $limitation): bool
```

File: code_samples/back_office/limitation/src/Security/MyPolicyProvider.php [hunk 12]
```diff
@@ -0,0 +1,15 @@
+<?php declare(strict_types=1);
+
+namespace App\Security;
+
+use Ibexa\Bundle\Core\DependencyInjection\Security\PolicyProvider\YamlPolicyProvider;
+
+class MyPolicyProvider extends YamlPolicyProvider
+{
+    protected function getFiles()
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply code_samples/back_office/limitation/src/Security/MyPolicyProvider.php:11]:
```suggestion
    protected function getFiles(): array
```

[the remark this answers is not in the document]
Comment (adriendupuis) [reply code_samples/back_office/limitation/src/Security/MyPolicyProvider.php:11]: Or
```suggestion
    /** @returns string[] */
    protected function getFiles(): array
```

File: code_samples/back_office/limitation/src/Security/Limitation/Mapper/CustomLimitationFormMapper.php [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        $form->add('limitationValues', CheckboxType::class, [
+            'label' => LimitationTranslationExtractor::identifierToLabel($data->getIdentifier()),
+            'required' => false,
+            'data' => $data->limitationValues['value'],
+            'property_path' => 'limitationValues[value]',
+        ]);
+    }
+
+    public function getFormTemplate()
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply code_samples/back_office/limitation/src/Security/Limitation/Mapper/CustomLimitationFormMapper.php:25]:
```suggestion
    public function getFormTemplate(): string
```

File: code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        if (!$value instanceof CustomLimitationValue) {
+            throw new InvalidArgumentException('$value', 'Must be of type: CustomLimitationValue');
+        }
+
+        if ($value->limitationValues['value']) {
+            return Type::ACCESS_GRANTED;
+        }
+
+        // If the limitation value is not set to `true`, then $currentUser, $object and/or $targets could be challenged to determine if the access is granted or not.
+        return Type::ACCESS_DENIED;
```
Comment (alongosz) [inline code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php:70]: For the case described in the inline comment, this should be ACCESS_ABSTAIN, but I think now we don't have a distinction between denied and abstain.

**Review by alongosz — APPROVED** [review]

File: code_samples/back_office/limitation/translations/ezplatform_content_forms_policies.en.yaml
Comment (adriendupuis) [inline code_samples/back_office/limitation/translations/ezplatform_content_forms_policies.en.yaml:1]: Should be ibexa_content_forms_policies.en.yaml

Comment (adriendupuis) [reply code_samples/back_office/limitation/translations/ezplatform_content_forms_policies.en.yaml:1]: Fixed in f89ce324b9ee0d67806c3a8e9ac75780202152dd

File: code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php (hunk 14 above)
Comment (adriendupuis) [inline code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php:70]: I guess this comment has somehow lost is role of placeholder for additional code.

The comment could also mention `Type::ACCESS_ABSTAIN` to have all cases.

```suggestion
        // If the limitation value is not set to `true`, then $currentUser, $object and/or $targets could be challenged to determine if the access is granted or not. Here or elsewhere. When passing the baton, a limitation can return Type::ACCESS_ABSTAIN
        return Type::ACCESS_DENIED;
```

Comment (adriendupuis) [reply code_samples/back_office/limitation/src/Security/Limitation/CustomLimitationType.php:70]: Discussion moved to https://github.com/ibexa/documentation-developer/pull/2195#discussion_r1516310529

**Review by mateuszbieniek — APPROVED** [review]

File: docs/permissions/custom_policies.md [hunk 15]
```diff
@@ -58,131 +60,246 @@ class MyPolicyProvider implements PolicyProviderInterface
...
+php bin/console translation:extract en --domain=forms --dir=src --output-dir=translations
 ```
 
-### Extending existing Policies
+### Existing Policies extension
 
-A `PolicyProvider` may provide new functions to a module, and additional Limitations to an existing function. 
-**It is however strongly encouraged to add functions to your own Policy modules.**
+A `PolicyProvider` may provide new functions to a module, and additional Limitations to an existing function.
+**It's however strongly encouraged to add functions to your own Policy modules.**
```
[the remark this answers is not in the document]
Comment (juskora) [reply docs/permissions/custom_policies.md:133]: And it should be put into a "note" block.

**Review by juskora — CHANGES_REQUESTED** [review]

**Review by juskora — APPROVED** [review]

**Pull request merged**